include/ofi_queue.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#ifndef _OFI_QUEUE_H_
+#define _OFI_QUEUE_H_
+
+#include "config.h"
+
+struct ofi_queue_node {
+	struct ofi_queue_node *next;
+};
+
+struct ofi_queue_root {
+	struct ofi_queue_node *head;
+	fastlock_t head_lock;
+
+	struct ofi_queue_node *tail;
+	fastlock_t tail_lock;
+
+	struct ofi_queue_node marker;
+};
+
+static inline
+struct ofi_queue_root *ofi_queue_alloc(void)
+{
+	struct ofi_queue_root *root =
+		calloc(1, sizeof(struct ofi_queue_root));
+	if (!root)
+		return NULL;
+
+	root->head = root->tail = &root->marker;
+
+	fastlock_init(&root->head_lock);
+	fastlock_init(&root->tail_lock);
+	
+	return root;
+}
+
+static inline
+void ofi_queue_push(struct ofi_queue_node *new_node, struct ofi_queue_root *root)
+{
+	new_node->next = NULL;
+	fastlock_acquire(&root->tail_lock);
+	root->tail->next = new_node;
+	root->tail = new_node;
+	fastlock_release(&root->tail_lock);
+}
+
+static inline
+struct ofi_queue_node *ofi_queue_pop(struct ofi_queue_root *root)
+{
+	struct ofi_queue_node *head, *next;
+
+	while (1) {
+		fastlock_acquire(&root->head_lock);
+		head = root->head;
+		next = head->next;
+		if (next == NULL) {
+			fastlock_release(&root->head_lock);
+			return NULL;
+		}
+		root->head = next;
+		fastlock_release(&root->head_lock);
+
+		if (head == &root->marker) {
+			ofi_queue_push(head, root);
+			continue;
+		}
+		return head;
+	}
```

Comment:
This is very inefficient when the average queue depth is close to 1.  Why is this complexity needed, versus just wrapping slist with a single lock?

Reply:
@shefty It seems that I found a real problem that cause issues with dlist_ts. I'm going to re-implement this PR in order to use dlist_ts. Are you OK with that?

Reply:
@shefty I remove the implementation of queue. Now CMD queue uses early implemented `dlist_ts`. But stil WIP, I need to address @a-ilango comments to split PR on separate commits